[buffer.reqmts] relax BidirectionalIterator requirements

vinniefalco opened the pull request:
This change adopts proposed wording from N4606 ([path.itr]/2) to relax the **BidirectionalIterator** requirements for buffer sequence iterators, allowing lvalues to be returned.

The wording from N4606 reads:

> A path::iterator is a constant iterator satisfying all the requirements of a bidirectional iterator (24.2.6)
> except that, for dereferenceable iterators a and b of type path::iterator with a == b, there is no requirement that *a and *b are bound to the same object. Its value_type is path.

jwakely commented:
As I said in #42:

> This belongs in a proposal to the committee, not as a github issue.

vinniefalco committed:
[buffer.reqmts] relax BidirectionalIterator requirements:

This removes the requirement for user-defined buffer sequence
iterators to obey the strict aliasing rule of ForwardIterator.

vinniefalco commented:
@jwakely Howard explained the process to me - I didn't know that Technical Specifications were closed for changes. I'll go through the proper procedure for filing an issue, and include my proposed language changes.

jwakely commented:
Thanks!

jwakely commented:
@vinniefalco you've made me realise I don't actually have an issues list for the Networking TS yet, everything else has been dealt with as a proposal paper or as an editorial change. I'll start a document to track issues, and add this as the first one (so consider the proper procedure completed!)

The committee can then process those issues next month at the next meeting.

vinniefalco commented:
The issue of iterator requirements in the Networking TS is now in the Active Issues List:
http://cplusplus.github.io/LWG/lwg-active.html#2779
